docs/adrs/0000-caching-dependencies.md

```diff
@@ -0,0 +1,59 @@
+# 0. Caching dependencies
+Date: 2021-05-21
+
+Status: Proposed
+
+# Context
+`actions/setup-node` is the 2nd most popular action in GitHub Actions. A lot of customers use it in conjunction with [actions/cache](https://github.com/actions/cache) to speed up dependencies installation.  
+See more examples on proper usage in [actions/cache documentation](https://github.com/actions/cache/blob/main/examples.md#node---npm).
+
+# Goals & Anti-Goals
+Integration of caching functionality into `actions/setup-node` action will bring the following benefits for action users:
+- Decrease the entry threshold for using the cache for Node.js dependencies and simplify initial configuration
+- Simplify YAML pipelines because no need additional steps to enable caching
+- More users will use cache for Node.js so more customers will have fast builds!
+
+As the first stage, we will add support for NPM dependencies caching. We can consider adding the same functionality for Yarn later.
+
+We don't persue the goal to provide wide customization of caching in scope of `actions/setup-node` action. The purpose of this integration is covering ~90% of basic use-cases. If user needs flexible customization, we should advice them to use `actions/cache` directly.
+
+# Decision
+- Add `cache` input parameter to `actions/setup-node`. For now, input will accept the following values: 
+  - `npm` - enable caching for npm dependencies
+  - `''` - disable caching (default value)
+  - Potentially, we will be able to extend this input to support Yarn
+- Cache feature will be disabled by default to make sure that we don't break existing customers. We will consider enabling cache by default in next major release (`v3`)
+- Add optional input `package-lock-path` that will allow to specify path to `package-lock.json` file path:
```

Comment:
@ahdbilal , I am not sure what is the best name for this input instead of `package-lock-path`.
The purpose of the input is specifying the path to file which hash will be used as caching key

Reply:
maybe `working-directory`. I will get others' opinions on it.

Reply:
This input should accept either directory path or file path so `working-directory` doesn't sound correct. Also, `working-directory` sounds very generic. It is not clear what should be put in this input

Reply:
agreed! let us get others' opinions on it.

Reply:
`cache-package-path` might be worth considering. For all the big setup actions, python, dotnet, node I think we would be primarily be focusing on speeding up caching of dependencies that in some way originate from **package managers**. For node that is `npm` and `yarn`, for Python this is `conda`, `pip`, for Dotnet this could be `nuget` or potentially `Chocolatey NuGet`.

I like having `cache` at the start because then it's immediately obvious that this is in some way related to caching and the earlier `cache` input. If we remove the `-lock` from this input then it becomes generic enough where we can reuse this across all of our first party actions.

`cache-dependency-path` is another option, but I don't like it that much. My vote goes to `cache-package-path`

Reply:
In my opinion, the providing ability to specify package lock path should be available in the first iteration. It is pretty difficult from implementation side and don't have underwater rocks.
Without this feature, we will drop all use-cases where package-lock.json / yarn.lock are located in any subfolder (that looks like popular case).

It is very very basic customization that "should be". Just my IMHO.

As for the naming of field, I would prefer `package-lock-path` because it is clearer what should be specified. But i am open for suggestions

Reply:
We have decided to not support custom location for `package-lock.json` and `yarn.lock` in the first iteration